src/QUICConnection.ts

```diff
@@ -410,73 +335,180 @@ class QUICConnection extends EventTarget {
+      // Which also depends on a mutual TLS handshake
+      // It is expected that multiple `recv` and `send` pairs
+      // will be called to complete the connection establishment
+      await Promise.race([
+        this.secureEstablishedP, // This might reject with a relevant error
+        abortP, // This might abort for some other reason!
+      ]);
+    } catch (e) {
+      let e_ = e;
+      // This should only be true if we are infact aborted due to start timeout
+      if (ctx.signal.aborted) {
+        // No `QUICStream` objects could have been created, however quiche stream
+        // state should be cleaned up, and this can be done synchronously
+        for (const streamId of this.conn.readable() as Iterable<StreamId>) {
+          this.conn.streamShutdown(streamId, quiche.Shutdown.Read, 0);
         }
-        await this.send(mon);
-      },
+        for (const streamId of this.conn.writable() as Iterable<StreamId>) {
+          this.conn.streamShutdown(streamId, quiche.Shutdown.Write, 0);
+        }
+        // According to RFC9000, closing while in the middle of a handshake
+        // should use a transport error code `APPLICATION_ERROR`.
+        // For this library we extend this "handshake" phase to include the
+        // the TLS handshake too.
+        // This is also the behaviour of quiche when the connection is not
+        // in a "safe" state to send application errors (where `app` parameter is true).
+        // https://www.rfc-editor.org/rfc/rfc9000.html#section-10.2.3-3
+        this.conn.close(
+          false,
+          ConnectionErrorCode.ApplicationError,
+          Buffer.from('')
+        );
+        const localError = this.conn.localError()!;
+        e_ = new errors.ErrorQUICConnectionLocal(
+          'Failed to start QUIC connection due to start timeout',
+          {
+            data: localError,
+            cause: e
+          }
+        );
+        this.rejectSecureEstablishedP(e_);
+        this.dispatchEvent(
+          new events.EventQUICConnectionError({
+            detail: e_
+          })
+        );
+        this.dispatchEvent(
+          new events.EventQUICConnectionClose({
+            detail: {
+              type: 'local',
+              ...localError
+            }
+          })
+        );
+        // Send out the close!
+        await this.send();
+      }
+      // Wait for the connection to be fully closed
+      // It is expected that max idle timer will eventually resolve this
+      await this.closedP;
+      // Throw the augmented exception if it is augmented
+      // Otherwise throw the original
+      throw e_;
+    } finally {
+      ctx.signal.removeEventListener('abort', abortHandler);
+    }
+    if (this.config.keepAliveIntervalTime != null) {
+      this.startKeepAliveIntervalTimer(this.config.keepAliveIntervalTime);
+    }
+    this.addEventListener(
+      events.EventQUICConnectionError.name,
+      this.handleEventQUICConnectionError,
+    );
```

Comment:
the `handleEventQUICConnectionError` is the main way of rejecting the `this.secureEstablishedP`. But it's only added to the listener *after*  awaiting `this.secureEstablishedP`. As a result, the `QUICConnection.start()` is not reacting to most connection errors including the idle timeout and is hanging.

Reply:
Yep, the general principle should be to:

* Add event listeners BEFORE async operations begin.
* Remove event listeners AFTER async operations finish.

This translates roughly to adding event listeners at the beginning of `start` and removing them at the end of the `stop`.

For `create` and `destroy`, it's sort of depends. Firstly you can only register handlers after the instance is constructed, otherwise you cannot access them. But also you would do at the end of destroy. However for dependencies that are asynchronously created it depends on the other of async operations. See `QUICClient.createQUICClient`, it's a bit complicated.

Reply:
This should be fixed by 5865ec86b261e00144901cf1869a16941e05e9de.